pkg/network/packet_source_linux.go

```diff
@@ -0,0 +1,145 @@
+// +build linux_bpf
+
+package network
+
+import (
+	"fmt"
+	"reflect"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/ebpf/manager"
+	"github.com/google/gopacket/afpacket"
+)
+
+var _ PacketSource = &packetSource{}
+
+// packetSource provides a RAW_SOCKET attached to an eBPF SOCKET_FILTER
+type packetSource struct {
+	*afpacket.TPacket
+	socketFilter *manager.Probe
+	socketFD     int
+
+	exit chan struct{}
+
+	// telemetry
+	polls     int64
+	processed int64
+	captured  int64
+	dropped   int64
+}
+
+func NewPacketSource(filter *manager.Probe) (*packetSource, error) {
+	rawSocket, err := afpacket.NewTPacket(
+		afpacket.OptPollTimeout(1*time.Second),
+		// This setup will require ~4Mb that is mmap'd into the process virtual space
+		// More information here: https://www.kernel.org/doc/Documentation/networking/packet_mmap.txt
+		afpacket.OptFrameSize(4096),
+		afpacket.OptBlockSize(4096*128),
+		afpacket.OptNumBlocks(8),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error creating raw socket: %s", err)
+	}
+
+	// The underlying socket file descriptor is private, hence the use of reflection
+	socketFD := int(reflect.ValueOf(rawSocket).Elem().FieldByName("fd").Int())
+
+	// Attaches DNS socket filter to the RAW_SOCKET
+	filter.SocketFD = socketFD
+	if err := filter.Attach(); err != nil {
+		return nil, fmt.Errorf("error attaching filter to socket: %s", err)
+	}
+
+	ps := &packetSource{
+		TPacket:      rawSocket,
+		socketFilter: filter,
+		socketFD:     socketFD,
+		exit:         make(chan struct{}),
+	}
+	go ps.pollStats()
+
+	return ps, nil
+}
+
+func (p *packetSource) Stats() map[string]int64 {
+	return map[string]int64{
+		"socket_polls":      atomic.LoadInt64(&p.polls),
+		"packets_processed": atomic.LoadInt64(&p.processed),
+		"packets_captured":  atomic.LoadInt64(&p.captured),
+		"packets_dropped":   atomic.LoadInt64(&p.dropped),
+	}
+}
+
+func (p *packetSource) VisitPackets(exit <-chan struct{}, visit func([]byte, time.Time) error) error {
+	for {
+		data, stats, err := p.ZeroCopyReadPacketData()
+
+		// Immediately retry for EAGAIN
+		if err == syscall.EAGAIN || err == afpacket.ErrTimeout {
```

Comment:
For `EAGAIN` should this not break out of the loop?

Reply:
Righto, done